include/seastar/http/client.hh

```diff
@@ -80,6 +80,7 @@ class connection : public enable_shared_from_this<connection> {
     // too and thus the connection will be persistent by default. If the server
     // responds with older version, this flag will be dropped (see recv_reply())
     bool _persistent = true;
+    unsigned _requests_submitted = 0;
```

Comment:
I'll be honest - I still don't understand why you need this count. Why doesn't your code know if it's a retry (which requires a new connection) or the original request? Why does it need to keep this flag as a counter inside the connection?

Reply:
OK, so first here's basically how a pool works:

```
    make_request(req):
        if (pool.empty())
            con = make_new_connection()
        else
            con = pool.pop_front()

        err = con.sen_request_and_recv_reply(req)

        if (!err)
            pool.push_back(con)
```

In simple words -- after successful serving a request (serving means sending the request, getting the reply and handling it) keep the connection in the pool so that next make_request could re-use already established connection rather than making a new one. It's, of course, a bit more complex as it takes into account keep-alive headers and maximum number of simultaneously opened connections, but basically it's like that.

Now, why do I need the counter. Let's assume that the retry is implemented in a straightforward manner, i.e. -- if con.send_request_and_recv_reply(req) failed just go ahead and make another attempt over new connection

```
    make_request(req):
        if (pool.empty())
            con = make_new_connection()
        else
            con = pool.pop_front()

        err = con.send_request_and_recv_reply(req)
        if (err) {
            con = make_new_connection()
            err = con.send_request_and_recv_reply(req)
        }

        if (!err)
            pool.push_back(con)
```

In that case if the pool was empty on entry, we effectively execute this sequence of steps:

```
            con = make_new_connection()
            err = con.send_request_and_recv_reply(req)
            if (err) {
                con = make_new_connection()
                err = con.send_request_and_recv_reply(req)
            }

```

but this makes little sense, because we already tried a request over new connection once, it failed, what's the point in trying it one more time?

Opposite to that, in case first attempt was made over connection from pool, retrying it has much more sense (more on that below) so the retrying code turns into

```
    make_request(req):
        if (pool.empty()) {
            con = make_new_connection()
            from_pool = false
        } else {
            con = pool.pop_front()
            from_pool = true
        }

        err = con.send_request_and_recv_reply(req)
        if (err && from_pool) {
            con = make_new_connection()
            err = con.send_request_and_recv_reply(req)
        }

        if (!err)
            pool.push_back(con)
```

But in case of http client code, carrying the from_pool boolean all the way from client::make_connection() is a bit problematic, it's simpler just to keep this "bit" on the connection itself. I could make it a `bool was_in_pool_ever` on the connection instead of a counter, but counting the number of requests served is potentially nice metrics, while that bool is not


Now why retrying a request after it failed over new connection doesn't make sense, and retrying it if it failed over connection from pool makes sense.

The thing is that from my experiments I see that AWS S3 server never closes its connections when serving the very 1st request over it. Maybe I ran the test for not long enough, but still -- it simply never happened. Often S3 server behaves in a polite way and sends back `connection: close` header in the reply to indicate that this request is fulfilled, but the server s no longer going to server more requests over it. And sometimes S3 server just closes the connection "in the middle of operation". From the client perspective this "in the middle" is two fold. First, the connection can get closed server-side while client is sending the request (header or body). This happens more often and I can explain it to myself that server just times-out the idling connection (idling, because it may have not yet seen the request on his side). Second, server can close connection while sending back the response. I can't explain this to myself, but it still happens.

Reply:
> And another physically broken scene is that server returned 503 and this can only be handled in the s3 client. 

Why? Doesn't http client sees 503 reply status?

> When 503 returned, retry should be made on a new connection with a different IP other than from the connection pool. 

Pool doesn't have IP at hand, it just has a bunch of sockets. And it's up to the client's factory whether all those sockets are connected to the same IP or different

Reply:
> Therefore, it is crucial to close idle connections promptly. 

Agree, but "promptly" is not something measurable. AFAIK client doesn't have any hint on whether to terminate idling connection. Neither it's good to create new connection for every request, keep-alive should work here too. And once it's ON, server closing its side in the middle of request can happen as I described in the PR top comment

Reply:
@xemul 
> but this makes little sense, because we already tried a request over new connection once, it failed, what's the point in trying it one more time?

Well, to me it makes every sense in the world :-( When some hardware in the server cluster fails, whatever connection we have might break in the middle. This can be some long-standing connection we took from a pool, or some shiny-new connection we initiated just 1 second ago - both can break. Why should we **not** retry the shiny-new connection that broke in the middle? It seems completely arbitrary to me.

What worries me about completely arbitrary rules is not the performance impact (like keeping an extra counter) but rather the code complexity - I think you are adding logic that a year from now (even a week from now) developers will not understand, and don't know how to test, and everything will just work the same without this logic, and the developer who wants to change something in the implementation won't know if it it's safe to change or not.

Reply:
> Why should we not retry the shiny-new connection that broke in the middle?

Because "shiny-new connection that broke in the middle" can be one of two -- this can be new connection that was created to serve the request because there were no ready to use connections in the pool (I'll call it A below), or this is the new connection that we created to retry current request, that failed once (I'll call it B below). The "why should we not retry" question applies to both. And there are two answers to it.

1. We don't retry the shiny-new connection
2. We do retry the shiny-new connection

You propose to go the 2nd route and in that case we should not only retry case B (above), but case A too, shouldn't we? And if we do, then we can retry infinitely, because A reproduces itself if connection fails. And if that's the case we should at some point stop retrying to avoid infinite loops.


Returning to my initial comment in this thread. We're converting the

```
    make_request(req):
        if (pool.empty())
            con = make_new_connection()
        else
            con = pool.pop_front()

        err = con.sen_request_and_recv_reply(req)

        if (!err)
            pool.push_back(con)
```

code into

```
    make_request(req):
        if (pool.empty()) {
            con = make_new_connection()
            from_pool = false
        } else {
            con = pool.pop_front()
            from_pool = true
        }

        err = con.send_request_and_recv_reply(req)
        if (err && from_pool) {
            con = make_new_connection()
            err = con.send_request_and_recv_reply(req)
        }

        if (!err)
            pool.push_back(con)
```

You find the `from_pool` concept excessive, I don't disagree with that, so we can go with the

```
    make_request(req):
        if (pool.empty())
            con = make_new_connection()
        else
            con = pool.pop_front()

        err = con.send_request_and_recv_reply(req)
        if (err) {
            con = make_new_connection()
            err = con.send_request_and_recv_reply(req)
        }

        if (!err)
            pool.push_back(con)
```

code.

In that case we might also apply "why don't we restart shiny-new connection" logic to the 2nd connection and make it look like

```
    make_request(req):
        attempts = N
again:
        if (pool.empty())
            con = make_new_connection()
        else
            con = pool.pop_front()

        err = con.send_request_and_recv_reply(req)
        if (err && --attempts > 0) {
            goto again;
        }

        if (!err)
            pool.push_back(con)
```

?

Reply:
I'm not sure I follow all the different options you're suggesting, or how it relates from what's in the code.
You told me earlier that you don't want to retry several times - just retry ONCE. I am not sure I fully understood why, but that was your decision and it does make some sense if the probability of a single failure is low (and you said this is why you don't want to reuse from the pool and start a new connection).

So, if you only retry ONCE, you can still do this retry no matter where you got the original connection from - if it was a recycled connection from the pool or a new connection because the pool happened to be empty.

Yes, you can start doing retries in a loop, and so on (and probably have some sort of limit on the number or duration of retries, etc.), but I think that's orthogonal to the question of whether or not to do the first (or only) retry.

Reply:
> And it's up to the client's factory whether all those sockets are connected to the same IP or different

That's true, but the client cannot decide to make requests on a new connection.

Reply:
@nyh , OK, maybe the tracking of per-connection requests and retry based on that was some overkill that doesn't really worth it. I removed it in the updated version, now retry just goes ahead and creates new connection

Reply:
This logic is removed, resolving the conversation